### ISSUE TITLE
Linux - use natively provided tools

### ISSUE THREAD

**Issue by Oxalin** [UniversalMediaServer/UniversalMediaServer#730]
On any OSes, the natively available tools should always be prefered to the one packaged in and installed by UMS (ffmpeg, ffmpeg64, tsMuxeR, tsMuxeR-new). The reason is that the user may have installed tools that are more up-to-date, providing better performance, new functionalities, corrected bugs, etc.

So, instead of looking at the tools provided by UMS, the first search should be with native tools.

**Comment by Oxalin** [comment]
Here is a patch for Linux. It introduces a new method called getNativeBinaryPath() which is responsible to identify is a given tool is natively provided by the OS.

This change should also be changed on Windows and OS X.

[0001-On-Linux-try-to-use-natively-installed-tools-instead.txt](https://github.com/UniversalMediaServer/UniversalMediaServer/files/35109/0001-On-Linux-try-to-use-natively-installed-tools-instead.txt)

**Comment by Oxalin** [comment]
@Nadahar Should I use another mean to propose patches? I could set up full git repo and use branches.

**Comment by Nadahar** [comment]
@Oxalin We usually prefer pull requests. Just fork UMS, optionally create a branch, push the changes and create a pull request. 

I'm not sure about the change though, there's a lot of trickery involved with parameters and compatibility tied to versions and modified builds involved here. While it might solve some issues to do it this way, it will very likely create new ones not using our bundled versions. It's easier to control what is happening when we know which versions of the tools that are in use. I'm not an expert in this field though, @SubJunk is probably the one that knows this best.

**Comment by valib** [comment]
@Oxalin agree with @Nadahar that using external tools we not tested may cause problems. You can use them and report to us when they are working well and we appreciate this. But we should decide to relese them or not.

**Comment by SubJunk** [comment]
Yeah like the others said, we avoid a lot of issues by testing versions of dependencies and in some cases we use custom versions of dependencies. FFmpeg, MEncoder and tsMuxeR in particular have had unstable pasts for various things, so whenever I update a dependency I run a series of tests to check that most of our use cases still work. If a custom version is used, it's possible and sometimes even likely that things will fail and then it will waste everyone's time trying to look for a bug that didn't need to exist.

**Comment by Oxalin** [comment]
I understand. Would you be opened to expose the path to the engines in the UI then? By default, it would use UMS' provided engines, but it could easily be overridden. It would expose variables available in the config file which are not yet.

**Comment by Nadahar** [comment]
I don't think what the others thing, but I think allowing overriding in the configuration file could be an idea. We could put a "disclaimer" above the parameters explaining the potential consequences. Exposing it in the GUI isn't such a great idea imo.

**Comment by SubJunk** [comment]
@Nadahar there are already those options in UMS.conf

@Oxalin I don't know if it's a good idea to put them in the GUI, I think settings like that are better in the config file since I think they will almost never be used

**Comment by Oxalin** [comment]
I must be on the other side of the spectrum about exposing options: I don't really like hidden options and variables. I prefer to have them exposed through something stating that this is an advanced option with a clear warning/explanation. Or under a kind of power user mode. Think of Chrome or Firefox if you want, where they both offer ways to use a GUI to edit advanced options.

Here is how I would see it. Let's say there is a check box to allow editing the path (manually typing it or browsing to it): if unchecked, UMS' provided engine is used. Or this could be the opposite: box checked to use provided engine, unchecked to allow editing the path. I would either group all these variables together (under Video Files Engines or as another submenu item in the tree) or I would had one per engine.

I'm willing to work on this as long as others are interested. Otherwise, I'll just drop this issue and live with it. I am opened to suggestion if you are interested.

**Comment by SubJunk** [comment]
@Oxalin if we used our web interface for program settings I would agree it's good to have lots in there, but considering we have the ugly Java interface which is already bursting with options... I think @Nadahar was working on a filter for setting difficulties so you can select them from a dropdown. That would be perfect for this.

**Comment by Sami32** [comment]
@Oxalin @Nadahar @SubJunk Could this issue be closed now ? Since all the answer's elements was already given i guess.
```
# ----------------------------------------------------------------------------
# Non-GUI Options
# ----------------------------------------------------------------------------

# ---< Binary tools paths >---------------------------------------------------
# Path to mencoder (absolute or relative from project.binaries.dir)
# Default:
#     Win: win32/mencoder.exe
#     Mac: osx/mencoder
#     Linux: mencoder + system PATH
mencoder_path =

# Path to ffmpeg (absolute or relative from project.binaries.dir)
# Default:
#     Win: win32/ffmpeg.exe
#     Mac: osx/ffmpeg
#     Linux: ffmpeg + system PATH
ffmpeg_path =

# Path to mplayer (absolute or relative from project.binaries.dir)
# Default:
#     Win: win32/mplayer.exe
#     Mac: osx/mplayer
#     Linux: mplayer + system PATH
mplayer_path =

# Path to tsMuxeR (absolute or relative from project.binaries.dir)
# Default:
#     Win: win32/tsMuxeR.exe
#     Mac: osx/tsMuxeR
#     Linux: tsMuxeR + system PATH
tsmuxer_path =

# Path to dcraw (absolute or relative from project.binaries.dir)
# Default:
#     Win: win32/dcrawMS.exe
#     Mac: osx/dcraw
#     Linux: dcraw + system PATH
dcraw_path =

# Path to FLAC (absolute or relative from project.binaries.dir)
# Default:
#     Win: win32/flac.exe
#     Mac: osx/flac
#     Linux: flac + system PATH
flac_path =

# Path to VideoLAN (absolute or relative from project.binaries.dir)
# Default:
#     Win: videolan/vlc.exe
#     Mac: /Applications/VLC.app/Contents/MacOS/VLC
#     Linux: vlc + system PATH
vlc_path =
```

**Comment by Oxalin** [comment]
Shouldn't this be seen as a feature request for a future release from the answers provided?
